### PR TITLE
Add support for CLI metadata and OVPs using JSON or file contents

### PR DESCRIPTION
**Pull request by kevin-bates**
Per [this comment](https://github.com/elyra-ai/elyra/issues/2376#issuecomment-1006798310) we have a sustainability issue with attempting to interpret complex schemas (nested object-valued properties, schema composition, references, etc.). This pull request attempts to alleviate those concerns while also better-positioning us with the ability to replicate metadata configurations by enabling the following...

- Mutually exclusive options `--file` and `--json` have been added where:
    -  `--file`: will have a value of a file name containing the metadata for the instance being created or updated.  The contents must be valid JSON and can contain the complete metadata (with a `metadata` stanza and top-order properties like `schema_name` and `display_name`) or, if `metadata` is not present, treated as _the_ `metadata` stanza (thereby allowing external applications to _inject_ their JSON into the metadata service without knowledge of the system-owned properties as they can also be specified on the command-line).
    - `--json`: will have a value of JSON string, containing the metadata for the instance being created or updated.  As with `--file` the contents of the string could be the complete metadata or only the `metadata` stanza itself.
- Overrides of values _within_ the contents of the `--file` or `--json` references via the command-line are also supported.  As a result, one can specify a different `--display_name` value than is in the referenced content, or a different `--name` value than what is inferred by the filename, enabling replication.  
- `--replace` functionality has been updated to apply individual property updates and no longer requires the complete instance definition.  (I could see this having utility, both in terms of being able to update complex schema types, but also in terms of security since, in theory, the contents can be protected by filesystem permissions, etc.)
- General object-valued properties are also supported, such that a given _property_ could be overridden via a file or JSON reference.  (Note that this is also true for _any_ top-level property (metadata or above)).  For example, if the schema defines an object-valued property (e.g., `authentication`) and a user wishes to replace one instance's authentication with a new value, the following command could be issued:
```
elyra-metadata install runtimes --schema_name=kfp --name=my_kubeflow_config --replace --authentication=/path/to/my/authentication_values.json
```
and replace instance `my_kubeflow_config`'s current authentication OVP with the contents of `/path/to/my/authentication_values.json`.
~(Note: As of this moment the above example doesn't work because we make _replacement_ (i.e., update), fully specify the object.  We should make `--replace` only allow the specification of the _changed_ data - along with the necessary data to identify the existing instance.  This PR should address that as well. See 3rd item.)~

Still to do:
- [x] Auto-detect complex schemas and warn that `--file` or `--json` options should be used
- [x] Allow for `--replace` to only specify the properties that will be changed - along with necessary identifying options.
- [x] Add tests for the new functionality
- [x] Update docs to include new options

This PR will remain in **draft** until all items have been completed.
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ptitzler** [comment]
A common [usability] issue with files is that it's not trivial to create them from scratch. Let's say, for example, a user wants to create a runtime configuration from a file. Is there something Elyra could do to  provide a "skeleton" that a user can use to create a file?

**Comment by kevin-bates** [comment]
> Is there something Elyra could do to provide a "skeleton" that a user can use to create a file?

Not really since the point of enabling `--file`, `--json`, or _JSON-rich object-valued properties_ is because the schema has become too complex - preventing Elyra from programmitcally producing such a template.  I think the common approach (i.e., 90% case) would be that users create instances using the UI and leverage those instances as a "template".  Any previously-created instances are guaranteed to meet minimum requirements (i.e., have the necessary required values).

I suppose some of the factory-based schemas could document a template, but that erodes when considering BYO schemas (in which the primary use of complex schemas exists - e.g., kfp and airflow authentication).

**Comment by akchinSTC** [comment]
possibly related : Export of metadata/configurations https://github.com/elyra-ai/elyra/issues/1066

**Comment by ptitzler** [comment]
> replace functionality has been updated to apply individual property updates and no longer requires the complete instance definition. (I could see this having utility, both in terms of being able to update complex schema types, but also in terms of security since, in theory, the contents can be protected by filesystem permissions, etc.)

Can you please also update the CLI examples/description in https://elyra.readthedocs.io/en/latest/user_guide/runtime-conf.html#managing-runtime-configurations-using-the-elyra-cli?

**Comment by kevin-bates** [comment]
Thanks @ptitzler - This exercise actually uncovered a bug where display-name was being required on updates.  This has been addressed (in the code and tests).  Here's a screenshot for the doc update:
image

File: docs/source/user_guide/command-line-interface.md [hunk 1]
```diff
@@ -77,6 +77,90 @@ Refer to the topics below for detailed information on how to use `elyra-metadata
  - [Manage runtime images](runtime-image-conf.html#managing-runtime-images-with-the-command-line-interface)
  - [Manage pipeline components](pipeline-components.html#managing-custom-components-using-the-elyra-cli)
 
+#### Creating and updating metadata with complex properties
+The `elyra-metadata` application derives its command-line options (aside from a handful of system options) directly from the schema associated with the referenced schemaspace. In most cases, the schema properties are straightforward and easily determined.  However, JSON schemas can also contain complex properties and references that are not within the scope of `elyra-metadata`.  This section presents ways to create and update instances built from complex schemas.
+
+Application-level properties within a schema reside as top-level proeprties within the schema's `metadata` stanza.  For example, here's the `code-snippet` schema for the `code-snippets` schemaspace:
```
Comment (kiersten-stokes) [inline docs/source/user_guide/command-line-interface.md:83]:
```suggestion
Application-level properties within a schema reside as top-level properties within the schema's `metadata` stanza. For example, here's the `code-snippet` schema for the `code-snippets` schemaspace:
```

**Review by kiersten-stokes — APPROVED** [review]
Just a small typo as noted below. Other than that, all the scenarios that I tested are working as expected!

File: docs/source/user_guide/runtime-conf.md [hunk 2]
```diff
@@ -90,6 +90,7 @@ To create a runtime configuration for a Kubeflow Pipelines deployment:
 
 ```bash
 elyra-metadata install runtimes \
+       --schema_name=kfp
```
Comment (ptitzler) [inline docs/source/user_guide/runtime-conf.md:93]: There's a continuation character missing:

```suggestion
       --schema_name=kfp \
```

File: docs/source/user_guide/runtime-conf.md [hunk 3]
```diff
@@ -101,36 +102,25 @@ elyra-metadata install runtimes \
...
        --api_password=mynewpassword \
-       --engine=Argo \
-       --cos_endpoint=http://minio-service.kubeflow:9000 \
-       --cos_auth_type="USER_CREDENTIALS" \
-       --cos_username=minio \
-       --cos_password=minio123 \
-       --cos_bucket=test-bucket \
-       --tags="['kfp', 'v1.1']" \
-       --schema_name=kfp
+       --tags="['kfp', 'v1.1']"
```
Comment (ptitzler) [inline docs/source/user_guide/runtime-conf.md:120]: As is, the command fails:
```
$ elyra-metadata install runtimes \
       --replace \
       --name="my_kubeflow_pipelines_runtime" \
       --schema_name=kfp \
       --api_password=mynewpassword \
       --tags="['kfp', 'v1.1']"
[I 2022-01-19 12:44:38,517.517] Upgrading runtime kfp instance 'my_kubeflow_pipelines_runtime' to include runtime_type 'KUBEFLOW_PIPELINES'...
The following exception occurred saving metadata instance for schema 'kfp': Username and password are not supported for the selected Kubeflow authentication type.
```

Comment (kevin-bates) [reply docs/source/user_guide/runtime-conf.md:120]: Thanks for catching this.  I can reproduce this and believe it is related to default value handling.  I need to take a closer look and submit a fix.

Comment (kevin-bates) [reply docs/source/user_guide/runtime-conf.md:120]: This issue is all in the CLI processing where options are built from the schema and properties with default values that were NOT also specified on the command line during replace operations were being included in the update payload.  Fix (and test update) forthcoming.

**Review by akchinSTC — APPROVED** [review]

**Review by ptitzler — APPROVED** [review]
LGTM

**Pull request merged**